src/Features/Core/Portable/IntroduceVariable/AbstractIntroduceParameterService.cs

```diff
@@ -0,0 +1,829 @@
+                    document, expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: false, cancellationToken));
+
+                if (!hasOptionalParameter && methodSymbol.MethodKind is not MethodKind.Constructor)
+                {
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_0, document, expression,
+                        methodSymbol, containingMethod, allOccurrences: false, trampoline: true, overload: false, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Extract_method_to_invoke_at_all_callsites_for_all_occurrences_of_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: true, overload: false, cancellationToken));
+
+                    actionsBuilder.Add(CreateNewCodeAction(
+                        FeaturesResources.Introduce_overload_with_new_parameter_for_0, document,
+                        expression, methodSymbol, containingMethod, allOccurrences: false, trampoline: false, overload: true, cancellationToken));
+                    actionsBuilder.Add(CreateNewCodeAction(
+                         FeaturesResources.Introduce_overload_with_new_parameter_for_all_occurrences_of_0, document,
+                         expression, methodSymbol, containingMethod, allOccurrences: true, trampoline: false, overload: true, cancellationToken));
+                }
+            }
+
+            return actionsBuilder.ToImmutable();
+
+            // Local function to create a code action with more ease
+            MyCodeAction CreateNewCodeAction(string actionName, Document document,
+                TExpressionSyntax expression, IMethodSymbol methodSymbol, SyntaxNode containingMethod,
+                bool allOccurrences, bool trampoline,
+                bool overload, CancellationToken cancellationToken)
+            {
+                return new MyCodeAction(actionName, c => IntroduceParameterAsync(
+                    document, expression, methodSymbol, containingMethod, allOccurrences, trampoline, overload, cancellationToken));
+            }
+        }
+
+        /// <summary>
+        /// Finds the matches of the expression within the same block.
+        /// </summary>
+        protected static async Task<IEnumerable<TExpressionSyntax>> FindMatchesAsync(Document originalDocument,
+            TExpressionSyntax expressionInOriginal, SyntaxNode withinNodeInCurrent,
+            bool allOccurrences, CancellationToken cancellationToken)
+        {
+            var syntaxFacts = originalDocument.GetRequiredLanguageService<ISyntaxFactsService>();
+            var originalSemanticModel = await originalDocument.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var result = new HashSet<TExpressionSyntax>();
```

Comment:
i don't know why this is a hashset.  nodes are already unique.

Reply:
never got used anywhere, just forgot to remove because, for some reaosn, some items don't get the "unused variable" lines underneath them